### ISSUE TITLE
[Bug]: Possible race conndition cause deploy error and forces replacement

### ISSUE THREAD

**Issue by hegerdes** [hetznercloud/terraform-provider-hcloud#734]
### What happened?

When running `terrafrom apply` with and x86 based vCPU system it can happen that the deployment fails with this error:

```
hcloud_server.worker_pool_arm64[0]: Creating...
hcloud_server.worker_pool_amd64[0]: Creating...
hcloud_server.control_plane_pool[0]: Creating...
hcloud_server.control_plane_pool[0]: Still creating... [10s elapsed]
hcloud_server.worker_pool_amd64[0]: Still creating... [10s elapsed]
hcloud_server.worker_pool_arm64[0]: Still creating... [10s elapsed]
hcloud_server.worker_pool_amd64[0]: Creation complete after 17s [id=35893547]
hcloud_server.control_plane_pool[0]: Still creating... [20s elapsed]
hcloud_server.worker_pool_arm64[0]: Still creating... [20s elapsed]
hcloud_server.worker_pool_arm64[0]: Creation complete after 28s [id=35893546]
hcloud_server.control_plane_pool[0]: Still creating... [30s elapsed]
╷
│ Error: hcloud/inlineAttachServerToNetwork: attach server to network: cannot perform operation because server is locked (locked)
│
│   with hcloud_server.control_plane_pool[0],
│   on infra-main.tf line 170, in resource "hcloud_server" "control_plane_pool":
│  170: resource "hcloud_server" "control_plane_pool" {
│
╵
```

If you rerun `terrafrom apply` it mostly works but triggers a forced replacement out of no reason:
```
Terraform will perform the following actions:

  # hcloud_server.control_plane_pool[0] is tainted, so must be replaced
-/+ resource "hcloud_server" "control_plane_pool" {
      + backup_window              = (known after apply)
      ~ datacenter                 = "fsn1-dc14" -> (known after apply)
      ~ firewall_ids               = [
          - 1012659,
          - 1012660,
        ] -> (known after apply)
      ~ id                         = "35893548" -> (known after apply)
      ~ ipv4_address               = "49.12.10.11" -> (known after apply)
      ~ ipv6_address               = "2a01:4f8:c17:2ec4::1" -> (known after apply)
      ~ ipv6_network               = "2a01:4f8:c17:2ec4::/64" -> (known after apply)
        name                       = "controlplane-node-amd64-suk79d"
      ~ status                     = "running" -> (known after apply)
        # (13 unchanged attributes hidden)

      + network {
          + alias_ips   = (known after apply)
          + ip          = "10.0.1.8"
          + mac_address = (known after apply)
          + network_id  = 3227623
        }

      - public_net {
          - ipv4         = 0 -> null
          - ipv4_enabled = true -> null
          - ipv6         = 0 -> null
          - ipv6_enabled = true -> null
        }
      + public_net {
          + ipv4         = (known after apply)
          + ipv4_enabled = true
          + ipv6         = (known after apply)
          + ipv6_enabled = true
        }
    }

Plan: 1 to add, 0 to change, 1 to destroy.
```

### What did you expect to happen?

The deployment goes through on the first try. 

### Please provide a minimal working example

My terraform code:

```hcl
terraform {
  required_version = ">= 1.0"
  required_providers {
    hcloud = {
      source  = "hetznercloud/hcloud"
      version = "1.42.0"
    }
    random = {
      source  = "hashicorp/random"
      version = "3.5.1"
    }
  }
}

# Create controlplane nodes
resource "hcloud_server" "control_plane_pool" {
  count       = var.control_plane_instance_count
  name        = "controlplane-node-amd64-${random_string.cp-rnd[count.index].result}"
  image       = var.node_image
  server_type = var.control_plane_instance_type
  location    = var.location
  user_data   = local.cloud_init_user_data
  labels = {
    managedby         = "terraform"
    k8s_control_plane = "true"
    k8s_master        = "true"
    k8s_cp            = "true"
    task              = "k8s"
    k8s               = "control-plane"
    arch              = "amd64"
    type              = var.control_plane_instance_type
  }
  ssh_keys           = [for key in hcloud_ssh_key.default : key.id]
  public_net {
    ipv4_enabled = var.per_instance_ipv4
    ipv6_enabled = true
  }
  network {
    network_id = hcloud_network.my_k8s_network.id
    ip         = "10.0.1.${count.index + 8}"
  }
  lifecycle {
    ignore_changes = [
      ssh_keys,
      user_data
    ]
  }
}
```

```json
{
  "control_plane_instance_count": 1,
  "control_plane_instance_type": "cpx31",
  "per_instance_ipv4": true
}
```
Version Info
```
Terraform v1.5.4
on windows_amd64
+ provider registry.terraform.io/hashicorp/azurerm v3.67.0
+ provider registry.terraform.io/hashicorp/local v2.4.0
+ provider registry.terraform.io/hashicorp/null v3.2.1
+ provider registry.terraform.io/hashicorp/random v3.5.1
+ provider registry.terraform.io/hetznercloud/hcloud v1.42.0

Your version of Terraform is out of date! The latest version
is 1.5.5. You can update by downloading from https://www.terraform.io/downloads.html
```

**Comment by apricote** [comment]
I have tested this >50 times locally and could not reproduce the issue.

Do you have any other resources in your terraform configuration that interact with servers? (Firewalls, Volumes, LBs, Network Attachements, Primary IPs, Floating IPs)?

It would also help if you could get debug logs when this happens. You can get debug logs by running `TF_LOG=TRACE TF_LOG_PATH=server-locked.log terraform apply`. You can then send them to me at `julian.toelle <at> hetzner-cloud.de` or post them here if you are sure that there is no sensitive information in the log.

**Comment by apricote** [comment]
The issue did actually occur after 63 retries. It happened to me with a Firewall that uses a label selector. This will create a `apply_firewall` action when the server is created and that action locks the server for some time.

We already have a `Retry` logic in place that retries the failing call 5 time. This logic is currently waiting for `try * 1 * time.Second` between the tries (0-4s) for a total wait time of [10 seconds](https://www.wolframalpha.com/input?i2d=true&i=Sum%5Bi%2C%7Bi%2C0%2C4%7D%5D). If the locking action takes longer than this, the issue occurs.

I will refactor the `Retry` logic to use an [exponential backoff](https://pkg.go.dev/github.com/hetznercloud/hcloud-go/hcloud#ExponentialBackoff). With the current default of 5 retries, this will wait for [31s](https://www.wolframalpha.com/input?i2d=true&i=Sum%5BPower%5B2%2Ci%5D%2C%7Bi%2C0%2C4%7D%5D) in total.

**Comment by hegerdes** [comment]
Thanks for the fast response and trying it 63 times 😲   
It happed once in a while but not consistently, but when it did it was annoying.

I do in fact have a `apply_firewall` with a label. I will try it again as soon as you publish a new release.

**Comment by apricote** [comment]
> Thanks for the fast response and trying it 63 times :astonished:

Got a script for that :D Its not the first flaky issue we have encountered here.

<details>
<summary><code>find-broken-action.sh</code></summary>

```sh
#!/usr/bin/env zsh

TAINT_RESOURCES=(
  "hcloud_server.tf_734"
)

recreate() {
  for res in $TAINT_RESOURCES ; do \
    terraform taint $res; \
  done;

  TF_LOG=TRACE TF_LOG_PATH=734.log terraform apply --auto-approve
}

while recreate; do echo "another happy landing"; done;
```

</details>

> I will try it again as soon as you publish a new release.

Released as [1.42.1](https://github.com/hetznercloud/terraform-provider-hcloud/releases/tag/v1.42.1) :rocket: Binaries will be available in a few minutes.
